### ISSUE TITLE
expose lilv_free() with features, for windows compatibility

### ISSUE THREAD

**Issue by x42** [drobilla/lilv#14]
Various functions e.g.  `abstract_path()` make the caller responsible for freeing the returned value with free().

On Windows this is not an option. Memory allocated by a library must be free()ed by that library/module

**Comment by x42** [comment]
N/M it's already there... sorry for the noise

**Comment by x42** [comment]
I was too quick. 

A plugin does not link against liblilv and cannot call lilv_free() directly.  A plugin receives abstract_path(), absolute_path() via the  LV2_State_Map_Path`feature.   Is `free()` needs to be part of the provided feature.

**Comment by drobilla** [comment]
See https://gitlab.com/lv2/lv2/tree/free-path and https://github.com/drobilla/lilv/commit/055e645099c66d2ac0394dbb8cc73e843ca7a2fd

**Comment by drobilla** [comment]
@x42 I've updated this on current master, and it seems fine as far as the (relatively limited) tests go.  Did you have a specific motivating case it would be good to check before merging this stuff?

**Comment by x42** [comment]
All state restore methods that use `map_path->abstract_path()`. I've #idef'ed out the `free()` call in many plugins like

https://github.com/x42/convoLV2/blob/ff291a3d5e60ddbcf9458c3ad24d4a06c8252acb/lv2.c#L473-L475

**Comment by drobilla** [comment]
I don't have a setup at the moment that can build stuff outside lv2kit and friends on Windows.  Let me rephrase: is there a specific motivating case you would like to check before this stuff gets merged? :)

**Comment by drobilla** [comment]
API-wise, as is so often the case, I really regret LV2 not having the ability to just jam stuff on the end of structs, getting a feature for this is pretty annoying, but... given reality, this seems fine to me.

**Comment by x42** [comment]
What is the envisaged way to test at compile-time if `LV2_STATE__freePath` is available?

**Comment by drobilla** [comment]
I didn't think about that.  The LV2 version I guess?

**Comment by drobilla** [comment]
You could also just keep the fallback ifdeffed out I suppose.

**Comment by x42** [comment]
is the URI guarenteed to be a `#define`? if so 
```
#ifdef LV2_STATE__freePath
...
#endif 
```
might work

**Comment by drobilla** [comment]
Same as anything else, really.  I guess having an LV2_VERSION declared in the header would be handy, but it's annoying to have generated stuff in the headers, and that's what build systems are for...

Anyway, I guess I can stick this stuff on master and it can get hammered out a bit in practice in Ardour before release since IIRC you aren't using releases anyway.

**Comment by x42** [comment]
> You could also just keep the fallback ifdeffed out as you did there I suppose.

I will do that for a while, at least until debian/stable picked up new LV2 headers.
Also `#ifdef LV2_STATE__freePath` looks like a reasonable test.

As far as I'm concerned, this issue can be closed. Thanks!

**Comment by drobilla** [comment]
Addressed in https://github.com/drobilla/lilv/commit/ac562cafe4c94db6965c4f715e276b2acc82bef0
